crates/mun_codegen/src/ir/body.rs

```diff
@@ -569,18 +565,21 @@ impl<'a, 'b, D: IrDatabase> BodyIrGenerator<'a, 'b, D> {
         }
     }
 
-    /// Given an expression and the type of the expression, optionally dereference the value.
-    fn opt_deref_value(&mut self, ty: hir::Ty, value: BasicValueEnum) -> BasicValueEnum {
+    /// Given an expression and its value optionally dereference the value to get to the actual
+    /// value. This is useful if we need to do an indirection to get to the actual value.
+    fn opt_deref_value(&mut self, expr: ExprId, value: BasicValueEnum) -> BasicValueEnum {
+        let ty = &self.infer[expr];
         match ty {
             hir::Ty::Apply(hir::ApplicationTy {
                 ctor: hir::TypeCtor::Struct(s),
                 ..
             }) => match s.data(self.db).memory_kind {
-                hir::StructMemoryKind::GC => deref_heap_value(&self.builder, value),
-                hir::StructMemoryKind::Value => value,
+                hir::StructMemoryKind::GC => return deref_heap_value(&self.builder, value),
+                hir::StructMemoryKind::Value => {}
             },
-            _ => value,
+            _ => {}
```

Comment:
what is the actual change here?

Reply:
First it returned values from the match, but I found it clearer to have it fall through if it didn't match so the default behavior would be more obvious (simply return the value).

Reply:
Maybe turn this into an `if let` statements in that case? That would remove quite some loc and shows that point a little better imo